analyses/hello-clusters/02_compare-clustering-parameters.Rmd

```diff
@@ -0,0 +1,466 @@
+In this section, we'll show an overview of how you might write code to vary two parameters at once (here, nearest neighbors and resolution as examples) and visualize results.
+
+
+```{r sweep two parameters}
+# Define vectors of parameters to vary
+nn_values <- seq(10, 30, 10)
+res_values <- seq(5, 15, 5) / 10
+
+
+cluster_results_list <- rOpenScPCA::sweep_clusters(
+  pca_matrix,
+  nn = nn_values,
+  resolution = res_values
+)
+```
+
+This resulting list now has 9 different clustering results, for each combination of `nn_values` and `res_values`:
+
+
+```{r length cluster_results_list two parameters}
+length(cluster_results_list)
+```
+
+
+### Visualize clusters
+
+Next, we'll iterate over `cluster_results_list` to plot the UMAPs.
+This time, we'll use `purrr::map()` and pull out parameters from each iteration's `cluster_df` to form the UMAP panel title.
+
+```{r plot nn res umaps, fig.height = 14}
+umap_plots <- cluster_results_list |>
+  purrr::map(
+    \(cluster_df) {
+      # Add a column with cluster assignments to umap_df
+      umap_df_plot <- umap_df |>
+        dplyr::mutate(cluster = cluster_df$cluster)
+
+      # Create a title for the UMAP with both parameters
+      umap_title <- glue::glue(
+        "nn: {unique(cluster_df$nn)}; res: {unique(cluster_df$resolution)}"
+      )
+
+      # Plot the UMAP, colored by the new cluster variable
+      ggplot(umap_df_plot, aes(x = UMAP1, y = UMAP2, color = cluster)) +
+        geom_point(alpha = 0.6) +
+        labs(title = umap_title) +
+        # We'll add a couple UMAP-specific plot settings
+        coord_equal() +
+        theme(
+          axis.ticks = element_blank(),
+          axis.text = element_blank(),
+          legend.position = "bottom"
+        )
+    }
+  )
+
+# Print the plots with patchwork::wrap_plots()
+patchwork::wrap_plots(umap_plots)
+```
+
+
+
+### Calculate and visualize QC metrics
+
+This section presents one coding strategy to calculate and visualize results when varying two clustering parameters.
+In particular, we use faceting to help display all information in one plot, by placing nearest neighbor values on the X-axis and faceting by resolution values.
+Since silhouette width and neighhorbood purity calculations using generally similar code, we'll just show neighborhood purity here.
+
+#### Neighborhood purity
+
+First, we'll calculate neighborhood purity and combine results into a single data frame.
+
+```{r calculate purity two parameters}
+purity_df <- cluster_results_list |>
+  purrr::map(
+    \(cluster_df) {
+      rOpenScPCA::calculate_purity(pca_matrix, cluster_df)
+    }
+  ) |>
+  dplyr::bind_rows()
+```
+
+
+We'll add a column `resolution_label` which we'll use to have informative panel titles in the faceted ggplot we make next.
+
+```{r add resolution_label column}
+purity_df <- purity_df |>
+  dplyr::mutate(resolution_label = glue::glue("Resolution: {resolution}"))
+```
```

Comment:
Not the biggest fan of this pattern. I might show off instead the `labeller` argument to `facet_wrap`